pkg/webhooks/pytorch/pytorchjob_webhook.go

```diff
@@ -0,0 +1,146 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pytorch
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"golang.org/x/exp/slices"
+
+	apimachineryvalidation "k8s.io/apimachinery/pkg/api/validation"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	trainingoperator "github.com/kubeflow/training-operator/pkg/apis/kubeflow.org/v1"
+)
+
+var (
+	specPath               = field.NewPath("spec")
+	pytorchReplicaSpecPath = specPath.Child("pytorchReplicaSpecs")
+)
+
+type Webhook struct{}
+
+func SetupWebhook(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&trainingoperator.PyTorchJob{}).
+		WithValidator(&Webhook{}).
+		Complete()
+}
+
+// +kubebuilder:webhook:path=/validate-kubeflow-org-v1-pytorchjob,mutating=false,failurePolicy=fail,sideEffects=None,groups=kubeflow.org,resources=pytorchjobs,verbs=create;update,versions=v1,name=validator.pytorchjob.training-operator.kubeflow.org,admissionReviewVersions=v1
+
+var _ webhook.CustomValidator = &Webhook{}
+
+func (w *Webhook) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	job := obj.(*trainingoperator.PyTorchJob)
+	log := ctrl.LoggerFrom(ctx).WithName("pytorchjob-webhook")
+	log.V(5).Info("Validating create", "pytorchJob", klog.KObj(job))
+	warnings, errs := validatePyTorchJob(job)
+	return warnings, errs.ToAggregate()
+}
+
+func (w *Webhook) ValidateUpdate(ctx context.Context, _ runtime.Object, newObj runtime.Object) (admission.Warnings, error) {
+	job := newObj.(*trainingoperator.PyTorchJob)
+	log := ctrl.LoggerFrom(ctx).WithName("pytorchjob-webhook")
+	log.V(5).Info("Validating update", "pytorchJob", klog.KObj(job))
+	warnings, errs := validatePyTorchJob(job)
+	return warnings, errs.ToAggregate()
+}
+
+func (w *Webhook) ValidateDelete(context.Context, runtime.Object) (admission.Warnings, error) {
+	return nil, nil
+}
+
+func validatePyTorchJob(job *trainingoperator.PyTorchJob) (admission.Warnings, field.ErrorList) {
+	var allErrs field.ErrorList
+	var warnings admission.Warnings
+
+	if errors := apimachineryvalidation.NameIsDNS1035Label(job.ObjectMeta.Name, false); len(errors) != 0 {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("metadata").Child("name"), job.Name, fmt.Sprintf("should match: %v", strings.Join(errors, ","))))
+	}
+	ws, err := validateSpec(job.Spec)
+	warnings = append(warnings, ws...)
+	allErrs = append(allErrs, err...)
+	return warnings, allErrs
```

Comment:
I guess, it's a new way of how KubeBuilder suggest to return errors/warnings: https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html, right ?
Rather than traceback error, they use ErrorList to append errors.

Maybe we modify our validation webhook for Katib as well sometime in the future: https://github.com/kubeflow/katib/blob/master/pkg/webhook/v1beta1/experiment/validation_webhook.go#L95

Reply:
> I guess, it's a new way of how KubeBuilder suggest to return errors/warnings: https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html, right ?
Rather than traceback error, they use ErrorList to append errors.

This is not a new way, this approach has been used in the kubernetes/kubernetes long term.

Anyway, I agree with applying this approach to Katib as well.

Reply:
Created an issue in the katib side: https://github.com/kubeflow/katib/issues/2302